src/auth/auth.service.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  BadRequestException,
+  ForbiddenException,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { PrismaService } from '../prisma/prisma.service';
+import { JwtService } from '@nestjs/jwt';
+import { LoginDto, RegisterDto } from './dto';
+import { Response } from 'express';
+import * as bcrypt from 'bcryptjs';
+import { v4 } from 'uuid';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private readonly prismaService: PrismaService,
+    private readonly jwtService: JwtService
+  ) {}
+
+  async register(registerDto: RegisterDto, res: Response) {
+    const { email, password, confirm_password } = registerDto;
+    const findUser = await this.prismaService.user.findUnique({ where: { email } });
+    if (findUser) {
+      throw new BadRequestException('This email is already in use! Please try again');
+    }
+    if (password !== confirm_password) {
+      throw new BadRequestException('Passwords do not match! Please try again');
+    }
+    const hashed_password: string = await bcrypt.hash(password, 12);
+
+    const newUser = await this.prismaService.user.create({
+      data: { email, password: hashed_password },
+    });
+    const tokens = await this.getTokens(newUser.id, newUser.email, newUser.role);
+    const hashedRefreshToken = await bcrypt.hash(tokens.refresh_token, 12);
+    const activationLink = v4();
+
+    await this.prismaService.user.update({
+      data: { hashedRefreshToken, activationLink },
+      where: { id: newUser.id },
+    });
+
+    res.cookie('refresh_token', tokens.refresh_token, {
+      maxAge: 7 * 24 * 60 * 60 * 1000,
+      httpOnly: true,
+    });
+    return tokens;
+  }
+  async login(loginDto: LoginDto, res: Response) {
+    const { email, password } = loginDto;
+    const user = await this.prismaService.user.findUnique({ where: { email } });
+    if (!user) throw new UnauthorizedException('User not found');
```

Comment:
move all this validation to a helper function like `validateUser`

Reply:
I updated all you said, Can you review one more time please?